### PR TITLE
CodeParrot data pretokenization

### PR DESCRIPTION
**Pull request by loubnabnl**
This PR adds code for data pretokenization of CodeParrot . In fact it takes long to tokenize the data especially for small models, so having a pretokenized dataset might improve the training speed. We also fix an error in the `README.md` and `scripts/initialize_model.py` inside the codeparrot repo. 

cc @lvwerra

### CODE REVIEW

File: examples/research_projects/codeparrot/scripts/pretokenizing.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_dir)
+
+t_start = time.time()
+ds = load_dataset(args.dataset_name, split="train")
+print(f"Dataset loaded in {time.time()-t_start:.2f}s")
+
+t_start = time.time()
+ds = ds.map(
+    tokenize,
+    num_proc=args.num_workers,
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/pretokenizing.py:30]: I am not sure parallelizing much on the Python side helps here a lot. The tokenizer already tokenizes in parallel efficiently in Rust and so passing batches to the `tokenize` function is probably more efficient.
```suggestion
    batched=True,
    batch_size=10_000,
```
Note that you need to change your `tokenize` function slightly since it will get batches instead if single samples.

File: examples/research_projects/codeparrot/scripts/pretokenizing.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+parser = HfArgumentParser(PretokenizationArguments)
+args = parser.parse_args()
+if args.num_workers is None:
+    args.num_workers = multiprocessing.cpu_count()
+tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_dir)
+
+t_start = time.time()
+ds = load_dataset(args.dataset_name, split="train")
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/pretokenizing.py:24]: It should be possible to stream the dataset here, too, right?

**Review by lvwerra — COMMENTED** [review]
Looks great! My main concern is about the parallelization of the tokenization. I think we can leave that up to the tokenizer which will handle it efficiently if we pass it large batches of text. Could you maybe run a quick test, e.g. on the validation set to see if it makes a difference?

File: examples/research_projects/codeparrot/scripts/pretokenizing.py (hunk 1 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/pretokenizing.py]: I made a test on the validation set, it seems like relying on the tokenizer and batches slows things, it takes 73s to tokenize this way while it takes 38s to tokenize using python multiprocessing (nvidia a100)

File: examples/research_projects/codeparrot/scripts/pretokenizing.py (hunk 2 above)
Comment (loubnabnl) [reply examples/research_projects/codeparrot/scripts/pretokenizing.py:24]: yes it should but when I activate streaming in the data loading the map() function doesn't have the argument remove_columns

File: examples/research_projects/codeparrot/README.md [hunk 3]
```diff
@@ -55,6 +55,16 @@ python scripts/preprocessing.py \
 ```
 During preprocessing the dataset is downloaded and stored locally as well as caches of the computations. Make sure you have more than 500GB free disk space to execute it.
 
+### Pretokenization
+The tokenization of the data might be slow during the training especially for small models. We provide code to pretokenize the data beforehand in `scripts/pretokenizing.py`, but we mention that this step is optional. The dataset is downloaded and stored locally but the tokenized data is directly pushed to the hub. The tokenized clean [train](https://huggingface.co/datasets/loubnabnl/tokenized-codeparrot-train) and [validation](https://huggingface.co/datasets/loubnabnl/tokenized-codeparrot-valid) datasets are available if you want to use them directly.
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:59]:
```suggestion
The tokenization of the data might be slow during the training especially for small models. We provide code to pretokenize the data beforehand in `scripts/pretokenizing.py`, but this step is optional. The dataset is downloaded and stored locally and the tokenized data is pushed to the hub. The tokenized clean [train](https://huggingface.co/datasets/loubnabnl/tokenized-codeparrot-train) and [validation](https://huggingface.co/datasets/loubnabnl/tokenized-codeparrot-valid) datasets are available if you want to use them directly.
```

File: examples/research_projects/codeparrot/README.md [hunk 4]
```diff
@@ -55,6 +55,16 @@ python scripts/preprocessing.py \
...
+
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:60]: Can you also add a note to the training section how to leverage the pretokenized data?

File: examples/research_projects/codeparrot/README.md [hunk 5]
```diff
@@ -55,6 +55,16 @@ python scripts/preprocessing.py \
...
+To execute the pretokenization, for the clean train data for instance, run the following command:
+```bash
+python scripts/pretokenizing.py \
+--dataset_name lvwerra/codeparrot-clean-train \
+--tokenized_data_repo <your Hugging Face hub username>/tokenized-codeparrot-train
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:65]: If no username is supplied then it always pushes to your own account. No need to do this explicitly.

```suggestion
--tokenized_data_repo tokenized-codeparrot-train
```

File: examples/research_projects/codeparrot/scripts/arguments.py [hunk 6]
```diff
@@ -172,6 +173,25 @@ class TokenizerTrainingArguments:
...
+
+    tokenizer_dir: Optional[str] = field(
+        default="lvwerra/codeparrot",
+        metadata={"help": "Name or path to the tokenizer."},
+    )
+    dataset_name: Optional[str] = field(
+        default="lvwerra/codeparrot-clean-train", metadata={"help": "Name or path to the dataset to pretokenize."}
+    )
+    tokenized_data_repo: Optional[str] = field(
+        default="lvwerra/tokenized-codeparrot-train", metadata={"help": "Repo name of the pretokenized data."}
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/scripts/arguments.py:190]:
```suggestion
        default="tokenized-codeparrot-train", metadata={"help": "Repo name of the pretokenized data."}
```

**Review by lvwerra — COMMENTED** [review]
Looks good to me, thanks for adding this! Left a few minor comments. If parallelization with `datasets` is faster then let's keep it.

File: examples/research_projects/codeparrot/README.md [hunk 7]
```diff
@@ -77,7 +87,8 @@ python scripts/initialize_model.py \
 ```
 This will initialize a new model with the architecture and configuration of `gpt2-large` and use the tokenizer to appropriately size the input embeddings. Finally, the initilaized model is pushed the the hub.
 
-Now that the dataset, tokenizer, and model are ready we can start training the model. The main training script is built with `accelerate` to scale across a wide range of platforms and infrastructure scales. We train two models with [110M](https://huggingface.co/lvwerra/codeparrot-small/) and [1.5B](https://huggingface.co/lvwerra/codeparrot/) parameters for 25-30B tokens on a 16xA100 (40GB) machine which takes 1 day and 1 week, respectively.
+For the dataset we can either use the preprocessed text data or the pretokenized data to speed up the training a little.
```
Comment (lvwerra) [inline examples/research_projects/codeparrot/README.md:90]:
```suggestion
We can either pass the name of a text dataset or a pretokenized dataset which speeds up training a bit.
```

**Review by lvwerra — APPROVED** [review]
One minor comment and then it is good to go 🚀
